### PR TITLE
WIP: nix builder for incremental compiling of spacchetti

### PR DESCRIPTION
**Pull request by Profpatsch**
This is not yet finished, but because some people have asked about it, here’s what I got so far.

The builder in `/nix/default.nix` is abstracted over the recursion of packages and what should be done with input sources can be influenced by writing small builder functions. There’s two of these functions in the module:

* `pscPackageSet` which creates a folder in the format used by `psc-package`.
* `compileAll`, which invokes the Purescript compiler on all inputs, with the correct dependencies. If some package changes, only the reverse dependencies are rebuilt.

`compileAll` is wired up at the moment, so if you run `nix-build nix/default.nix` from the `spacchetti` root dir, it will compile all purescript packages in the repository.

### CODE REVIEW

File: git-sha.json [hunk 1]
```diff
@@ -0,0 +1 @@
...unordered-collections": "09rrsym4w0xainsxzgd1ma6ay4iyx2kbb3x3w4k2bp552bq3vmj1", "unsafe-coerce": "0k9255mk2mz6xjb11pwkgfcblmmyvr86ig5kr92jwy95xim09zip", "unsafe-reference": "0q758dz59qz0li4s3w1qcg921xp5i5rh6i1l611iv7rr8cbj11al", "validation": "1y5i1jpr6jn1yjjpi4l5mk7jyd5v21fym892s0y5bdn4nj7i465l", "variant": "1awxcl3kc3z3hm8kxdg0w74v76pj636nk31d4ggqh49y25v3mpxj", "web-clipboard": "0zbhrm4ck8a8wf1knpmxccfyws07frld8i70wnkwjfan1pgdji63", "web-dom": "070ybc5xzpsh75p618ll9j0lzxqgvc4cl5cdrn71v0vagmyhgm8h", "web-events": "0hymk08prp29dm8v3xq079m7inkspb6gsyyynwqrcxsl3mha1f51", "web-file": "14kzwrwbfacxr4krfriy8qx13wnhiydrgs6hp59a7bmx58w0ifks", "web-html": "1y0g2ginqmfng058c99dm9n25vhz3g76d4l2v7yyrln9w04m4mvx", "web-socket": "1dld6dvwyhpxwcj28grnndjbamdfmy2imn7sl2824n74s03sy9kj", "web-storage": "1lr4lswsp62kl21rl6jb377a0sya105sfimkxm3y22grx88l6vj1", "web-touchevents": "1avg7jljd1j7d0b7k7fbqxi90aa01awp9x17v53mynzg6gyvc8zb", "web-uievents": "1683ysabddnhsjiy70axpslagi2cp2ab3qd0r1qjx6ab2zjd6kc0", "web-xhr": "05s3wp525wyvfz3rv456shvng9c0a4ncjdc7yxn2zglhq973km23", "xiaomian": "1n2ghz1z04xqxmjqdh0phnslq6fc7027j70si0lbp7njic6xljjp", "yargs": "10xbiyghplg54n9p7vs7mba9gmqfsbyiz0045c5p4p3v6djj60w3" }
```
Comment (boothead) [inline git-sha.json]: Does this change often? Diffs are going to be hard to spot on one line like this, so if it's generated is there a way to have pretty printed?

Comment (Profpatsch) [reply git-sha.json]: The file shouldn’t be checked in in the final version. Just to make it easy to try it out for now.

Comment (Profpatsch) [reply git-sha.json]: I don’t know whether I’d like to keep the perl script to be honest, we already have some Haskell setup.

**Comment by justinwoo** [comment]
Should we merge this? Does it work well enough for you?

The only thing is that because of externs files and how recompilation checks are done in the PureScript compiler currently (unknown if this would change anytime in the future), precompiling packages just won't work.

**Comment by justinwoo** [comment]
Closing due to out of scope

**Pull request closed without merging**